// @metaplex-foundation/mpl-core@1.0.1/dist/src/plugins/dataStore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dataStoreManifest = exports.dataStoreFromBase = exports.dataStoreUpdateInfoArgsToBase = exports.dataStoreInitInfoArgsToBase = void 0;
const lib_1 = require("./lib");
const pluginAuthority_1 = require("./pluginAuthority");
function dataStoreInitInfoArgsToBase(d) {
    return {
        dataAuthority: (0, pluginAuthority_1.pluginAuthorityToBase)(d.dataAuthority),
        initPluginAuthority: d.initPluginAuthority
            ? (0, pluginAuthority_1.pluginAuthorityToBase)(d.initPluginAuthority)
            : null,
        schema: d.schema ? d.schema : null,
    };
}
exports.dataStoreInitInfoArgsToBase = dataStoreInitInfoArgsToBase;
function dataStoreUpdateInfoArgsToBase(d) {
    return {
        schema: d.schema ? d.schema : null,
    };
}
exports.dataStoreUpdateInfoArgsToBase = dataStoreUpdateInfoArgsToBase;
function dataStoreFromBase(s, r, account) {
    return {
        ...s,
        dataAuthority: (0, pluginAuthority_1.pluginAuthorityFromBase)(s.dataAuthority),
        data: (0, lib_1.parseExternalPluginAdapterData)(s, r, account),
    };
}
exports.dataStoreFromBase = dataStoreFromBase;
exports.dataStoreManifest = {
    type: 'DataStore',
    fromBase: dataStoreFromBase,
    initToBase: dataStoreInitInfoArgsToBase,
    updateToBase: dataStoreUpdateInfoArgsToBase,
};
