// @metaplex-foundation/mpl-core-das@0.0.2/dist/src/das.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.das = void 0;
const mpl_core_1 = require("@metaplex-foundation/mpl-core");
const constants_1 = require("./constants");
const helpers_1 = require("./helpers");
async function searchAssets(context, input) {
    const dasAssets = await context.rpc.searchAssets({
        ...input,
        interface: (input.interface ?? constants_1.MPL_CORE_ASSET),
        burnt: false,
    });
    const mappedAssets = dasAssets.items.map((dasAsset) => (0, helpers_1.dasAssetToCoreAssetOrCollection)(dasAsset));
    if (input.interface === constants_1.MPL_CORE_COLLECTION || input.skipDerivePlugins) {
        return mappedAssets;
    }
    return (0, mpl_core_1.deriveAssetPluginsWithFetch)(context, mappedAssets);
}
function searchCollections(context, input) {
    return searchAssets(context, { ...input, interface: constants_1.MPL_CORE_COLLECTION });
}
function getAssetsByOwner(context, input) {
    return searchAssets(context, {
        ...input,
        owner: input.owner,
    });
}
function getAssetsByAuthority(context, input) {
    return searchAssets(context, {
        ...input,
        authority: input.authority,
    });
}
function getAssetsByCollection(context, input) {
    return searchAssets(context, {
        ...input,
        grouping: ['collection', input.collection],
    });
}
function getCollectionsByUpdateAuthority(context, input) {
    return searchCollections(context, {
        ...input,
        authority: input.updateAuthority,
    });
}
exports.das = {
    searchAssets,
    searchCollections,
    getAssetsByOwner,
    getAssetsByAuthority,
    getAssetsByCollection,
    getCollectionsByUpdateAuthority,
};
